### PR TITLE
Add styx-client module

### PR DESCRIPTION
**Pull request by fabriziodemaria**
The CLI would use the `styx-client` to perform the requests, also part of this PR.

**TODO**
- The CLI version is part of the requests' header; (removed)

_NOTE_
I am not sure about the way the client handles errors and exceptions at this point. The `throwable` we add to a failing `CompletionStage` is different in different cases but very obscure to users. 
I think we should extract at least two kind of checked exceptions: `ClientErrorException` and `ServerErrorException`. A more generic `IOException` can also be added to the signatures for unexpected arguments/payloads. In this way errors can be better documented and easier to handle.
However I am not sure what are the best practices for this case.

### CODE REVIEW

File: styx-cli/src/main/java/com/spotify/styx/cli/Main.java [hunk 1]
```diff
@@ -250,265 +232,158 @@ private void run() {
...
-      try {
-        Event typedEvent = OBJECT_MAPPER.convertValue(event, Event.class);
-        eventName = EventUtil.name(typedEvent);
-        eventInfo = CliUtil.info(typedEvent);
-      } catch (IllegalArgumentException e) {
-        // fall back to just inspecting the json
-        eventName = event.get("@type").asText();
-        eventInfo = "";
-      }
+    workflowFuture.thenCombine(workflowStateFuture, Tuple::of)
```
Comment (fabriziodemaria) [inline styx-cli/src/main/java/com/spotify/styx/cli/Main.java]: I am looking for a way to achieve the same without using `Tuple`...

File: styx-cli/src/main/java/com/spotify/styx/cli/Main.java [hunk 2]
```diff
@@ -90,9 +60,9 @@
   private static final String PARAMETER_DEST = "parameter";
 
   private static final int EXIT_CODE_SUCCESS = 0;
-  private static final int EXIT_CODE_UNKNOWN_ERROR = 1;
-  private static final int EXIT_CODE_ARGUMENT_ERROR = 2;
-  private static final int EXIT_CODE_API_ERROR = 3;
+  private static final int EXIT_CODE_ARGUMENT_ERROR = 1;
```
Comment (bergman) [inline styx-cli/src/main/java/com/spotify/styx/cli/Main.java]: I think 1 for unknown is nicer since whenever i see something other than 1 I assume that there was a specific error.

File: styx-cli/pom.xml [hunk 3]
```diff
@@ -69,6 +69,11 @@
       <artifactId>jackson-annotations</artifactId>
       <version>${jackson.version}</version>
     </dependency>
+      <dependency>
+          <groupId>com.spotify</groupId>
+          <artifactId>styx-client</artifactId>
+          <version>1.0.6-SNAPSHOT</version>
```
Comment (honnix) [inline styx-cli/pom.xml]: We should probably starting using `<version>${project.version}</version>` for inter-module dependency. Or even move to `dependencyManagement`. Maybe a separated PR. However things like `<version>${jackson.version}</version>` in this pom.xml should be gone.

File: styx-cli/src/main/java/com/spotify/styx/cli/Main.java (hunk 2 above)
Comment (honnix) [inline styx-cli/src/main/java/com/spotify/styx/cli/Main.java]: Agree.

File: styx-cli/pom.xml (hunk 3 above)
Comment (TinaRanic) [inline styx-cli/pom.xml]: yeah, the versions for some of the dependencies, like argparse4j and jackson are already defined in the parent pom under deps management, so no need to specify them here.

File: styx-cli/src/main/java/com/spotify/styx/cli/Main.java [hunk 4]
```diff
@@ -250,265 +232,158 @@ private void run() {
...
-    final BackfillInput backfillInput = BackfillInput.create(
-        Instant.parse(start), Instant.parse(end), component, workflow, concurrency);
-
-    final ByteString payload = serialize(backfillInput);
-    final ByteString response = send(
-        Request.forUri(apiUrl("backfills"), "POST").withPayload(payload));
-    final Backfill backfill = deserialize(response, Backfill.class);
-    cliOutput.printBackfill(backfill);
+    styxClient.backfillCreate(component, workflow, start, end, concurrency)
+        .whenComplete((backfill, throwable) -> cliOutput.printBackfill(backfill));
```
Comment (honnix) [inline styx-cli/src/main/java/com/spotify/styx/cli/Main.java]: Shouldn't we check throwable as well here? Since in the exception handler we don't really use response, `CompletionStage:exceptionally` would be better?

File: styx-cli/src/main/java/com/spotify/styx/cli/Main.java [hunk 5]
```diff
@@ -250,265 +232,158 @@ private void run() {
...
-      editedBackfillBuilder.concurrency(concurrency);
-    }
-    final ByteString putPayload = serialize(editedBackfillBuilder.build());
-    final ByteString putResponse = send(
-        Request.forUri(apiUrl("backfills", id), "PUT").withPayload(putPayload));
-
-    final Backfill newBackfill = deserialize(putResponse, Backfill.class);
-    cliOutput.printBackfill(newBackfill);
+    styxClient.backfillEditConcurrency(id, concurrency)
+        .whenComplete((backfill, throwable) -> cliOutput.printBackfill(backfill));
```
Comment (honnix) [inline styx-cli/src/main/java/com/spotify/styx/cli/Main.java]: Same here, exception handling.

File: styx-cli/src/main/java/com/spotify/styx/cli/Main.java [hunk 6]
```diff
@@ -250,265 +232,158 @@ private void run() {
...
   }
 
-  private void backfillHalt() throws ExecutionException, InterruptedException {
+  private void backfillHalt() {
     final String id = namespace.getString(parser.backfillHaltId.getDest());
-    send(Request.forUri(apiUrl("backfills", id), "DELETE"));
+
+    styxClient.backfillHalt(id).whenComplete((none, throwable) -> cliOutput.printMessage(
```
Comment (honnix) [inline styx-cli/src/main/java/com/spotify/styx/cli/Main.java]: exception handling

File: styx-cli/src/main/java/com/spotify/styx/cli/Main.java (hunk 1 above)
Comment (honnix) [inline styx-cli/src/main/java/com/spotify/styx/cli/Main.java]: `Tuple` is not bad in this case.

File: styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+                          final String apiHost,
+                          final String userAgentIdentifier) {
+    this.client = client;
+    this.apiHost = apiHost;
+    this.userAgentIdentifier = userAgentIdentifier;
+  }
+
+  @Override
+  public CompletionStage<RunStateDataPayload> activeStates(Optional<String> componentId) {
+    String url = apiUrl("status", "activeStates");
```
Comment (honnix) [inline styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java]: This kind of URL handling feels quite primitive, we should probably switch to https://square.github.io/okhttp/3.x/okhttp/okhttp3/HttpUrl.html. Have a `TODO` maybe? Not to get scope creep.

File: styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+            final JsonNode event = eventWithTimestamp.get("event");
+
+            String eventName;
+            String eventInfo;
+            try {
+              Event typedEvent = OBJECT_MAPPER.convertValue(event, Event.class);
+              eventName = EventUtil.name(typedEvent);
+              eventInfo = EventUtil.info(typedEvent);
+            } catch (IllegalArgumentException e) {
+              // fall back to just inspecting the json
```
Comment (bergman) [inline styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java]: Should the client be this relaxed with the types? The CLI was made that way because people don't upgrade right away but I'm not sure that should apply to all clients.

File: styx-cli/src/main/java/com/spotify/styx/cli/Main.java (hunk 2 above)
Comment (fabriziodemaria) [inline styx-cli/src/main/java/com/spotify/styx/cli/Main.java]: I was trying to follow an "order by severity" numbering, but reverted back to the original solution.

File: styx-cli/src/main/java/com/spotify/styx/cli/Main.java (hunk 4 above)
Comment (fabriziodemaria) [inline styx-cli/src/main/java/com/spotify/styx/cli/Main.java]: The problem with `exceptionally` is that it accepts a `Function` that must return a value, so it gets a bit messy with `return null` and explicit casts to make it work. Probably worse than the unused `response` in the `BiConsumer` of `whenComplete`.

**Comment by honnix** [comment]
Overall looks very good!

One thing from the users' point of view. How would a user instantiate `StyxApolloClient`? `StyxApolloClient client = new StyxApolloClient()` feels a bit odd.

How about something like this? Maybe not make much sense.

```
interface StyxClient extends StyxWorkflowClient, StyxBackfillClient, StyxResourceClient,
                             StyxSchedulerClient, StyxStatusClient {
}

class StyxApolloClient implements StyxClient {
}

StyxClient styxClient = StyxClient.builder()
    .client(client)
    .apiHost(apiHost)
    .userAgentIdentifier(userAgentIdentifier)
    .build();

StyxBackfillClient styxClient = StyxClient.builder()
    .client(client)
    .apiHost(apiHost)
    .userAgentIdentifier(userAgentIdentifier)
    .buildBackfillClient(); // just a cast here
```

File: styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+  private final Client client;
+  private final String apiHost;
+  private final String userAgentIdentifier;
+
+  public StyxApolloClient(final Client client,
+                          final String apiHost,
+                          final String userAgentIdentifier) {
+    this.client = client;
+    this.apiHost = apiHost;
+    this.userAgentIdentifier = userAgentIdentifier;
```
Comment (honnix) [inline styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java]: Hmm, this should be part of the client now, because what we want is to block the client, not cli.

Comment (honnix) [inline styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java]: `this.userAgentIdentifier = StyxApolloClient.class.getPackage().getImplementationVersion();`

**Comment by honnix** [comment]
If for sure we can differentiate client and server error, it makes sense to have specific exceptions. Then users can catch and see the problem.

File: styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+    final String url = apiUrl("backfills", backfillId);
+    return executeRequest(Request.forUri(url), BackfillPayload.class);
+  }
+
+  @Override
+  public CompletionStage<BackfillsPayload> backfillList(Optional<String> componentId,
+                                                        Optional<String> workflowId,
+                                                        boolean showAll,
+                                                        boolean status) {
+    List<String> queries = new ArrayList<>();
```
Comment (honnix) [inline styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java]: final

File: styx-cli/src/main/java/com/spotify/styx/cli/Main.java [hunk 11]
```diff
@@ -253,262 +235,161 @@ private void run() {
...
-        .payload().orElse(ByteString.EMPTY);
-    final ByteString statePayload = requireSuccess(stateResponse.get())
-        .payload().orElse(ByteString.EMPTY);
+  private void workflowShow() throws ExecutionException, InterruptedException {
+    final String component = namespace.getString(parser.workflowShowComponentId.getDest());
+    final String workflow = namespace.getString(parser.workflowShowWorkflowId.getDest());
 
-    final Workflow workflow = deserialize(workflowPayload, Workflow.class);
-    final WorkflowState workflowState = deserialize(statePayload, WorkflowState.class);
+    CompletableFuture<Workflow> workflowFuture =
```
Comment (honnix) [inline styx-cli/src/main/java/com/spotify/styx/cli/Main.java]: final

File: styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+          return response;
+        default:
+          final String message = response.status().code() + " " + response.status().reasonPhrase();
+          throw new ApiErrorException(message, response.status().code());
+      }
+    });
+  }
+
+  private String apiUrl(String... parts) {
+    List<String> encodedPartsList = Arrays.stream(parts).map(part -> {
```
Comment (honnix) [inline styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java]: final

File: styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+        return URLEncoder.encode(part, UTF_8);
+      } catch (UnsupportedEncodingException e) {
+        throw new RuntimeException(e);
+      }
+    }).collect(Collectors.toList());
+    return "http://" + apiHost + STYX_API_ENDPOINT + "/" + String.join("/", encodedPartsList);
+  }
+
+  private String addQueryToApiUrl(String url, List<String> queries) {
+    List<String> encodedQueryParts = queries.stream().map(query -> {
```
Comment (honnix) [inline styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java]: final

**Comment by honnix** [comment]
LGTM apart from a few `final`s. Didn't check all actual logic. I assume you have tested against styx service. Would be nicer if we can add test cases later in another PR. Regarding that URL handling comment, I would still prefer adding a TODO and later using the offering from okhttp.

File: styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java (hunk 8 above)
Comment (fabriziodemaria) [inline styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java]: I think it is fine to be relaxed in the client for this API as well, since `EventInfo` is mainly "informational" thus it feels ok to make a best effort in constructing it. This detail about forward compatibility can be mentioned in the Javadocs though, I will add it.

**Comment by bergman** [comment]
👍
